core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeInMemorySorter.java

```diff
@@ -333,11 +333,12 @@ public UnsafeSorterIterator getSortedIterator() {
     if (nullBoundaryPos > 0) {
       assert radixSortSupport != null : "Nulls are only stored separately with radix sort";
       LinkedList<UnsafeSorterIterator> queue = new LinkedList<>();
-      if (radixSortSupport.sortDescending()) {
-        // Nulls are smaller than non-nulls
+
+      // The null order is either LAST or FIRST, regardless of sorting direction (ASC|DESC)
+      if (radixSortSupport.nullOrder() == PrefixComparator.NullOrder.LAST) {
         queue.add(new SortedIterator((pos - nullBoundaryPos) / 2, offset));
         queue.add(new SortedIterator(nullBoundaryPos / 2, 0));
-      } else {
+      } else if (radixSortSupport.nullOrder() == PrefixComparator.NullOrder.FIRST) {
         queue.add(new SortedIterator(nullBoundaryPos / 2, 0));
         queue.add(new SortedIterator((pos - nullBoundaryPos) / 2, offset));
       }
```

Comment:
I think one of the reasons I need to modify the `PrefixComparators` code is because` `UnsafeInMemorySorter` here puts null rows in a separate queue that is appended to before or after the non-null queue depending on sort direction ASC|DESC.